options.go

```diff
@@ -0,0 +1,67 @@
+package proteus
+
+import (
+	"io"
+
+	"github.com/simplesurance/proteus/sources"
+)
+
+// Option allows specifying options when creating a configuration parser.
+type Option func(*settings)
+
+type settings struct {
+	sources           []sources.Source
+	autoUsageExitFn   func()
+	autoUsageHeadline string
+	autoUsageWriter   io.Writer
+	loggerFn          Logger
+}
+
+func (s *settings) apply(options ...Option) {
+	for _, o := range options {
+		o(s)
+	}
+}
+
+// WithSources adds configuration sources to be used to read values for the
+// application parameters. It can be called multiple times. If a parameter is
+// found in more than one source, the source that was added first takes
+// precedence.
+func WithSources(s ...sources.Source) Option {
+	return func(p *settings) {
+		p.sources = append(p.sources, s...)
+	}
+}
+
+// WithAutoUsage will parse the `--help` command-line flag. If it is present,
+// will show usage and exit using the provided exitFn. If the exitFn is nil
+// it exists with os.Exit(0).
+func WithAutoUsage(writer io.Writer, headline string, exitFn func()) Option {
```

Comment:
I love the suggestion of having auto-usage enabled by default with some sane settings, code is updated to have it.

I don't like much the idea of having one option for `exitFn` and another for writer. The `exitFn` used in one purpose may not serve another purpose (for example, exiting with status 0 vs 1), and same thing for the writer.

The third point, on using gnu-style, what you suggested fits the `UsageErr` case. For example:

```bash
mkdir --ni
# Output:
# mkdir: unrecognized option '--ni'
# Try 'mkdir --help' for more information.
```

In case of the user asking for usage help:

```bash
mkdir --help
# Output:
# Usage: mkdir [OPTION]... DIRECTORY...
# Create the DIRECTORY(ies), if they do not already exist.
# 
# Mandatory arguments to long options are mandatory for short options too.
#  -m, --mode=MODE   set file mode (as in chmod), not a=rwx - umask
```

I took heavy inspiration on that, but made a few changes only where it looked necessary. Please check the updated code.